Mathlib/NumberTheory/Ostrowski.lean

```diff
@@ -55,4 +57,213 @@ lemma equiv_on_Nat_iff_equiv : (∃ c : ℝ, 0 < c ∧ (∀ n : ℕ , (f n)^c =
+
+/-- There exists a minimal positive integer with absolute value smaller than 1 -/
+lemma exists_smallest_Nat_zero_lt_mulRingNorm_lt_one : ∃ (p : ℕ), (0 < f p ∧ f p < 1) ∧
+    ∀ (m : ℕ), 0 < f m ∧ f m < 1 → p ≤ m := by
+  -- There is a positive integer with absolute value different from one
+  have hn : ∃ (n : ℕ), n ≠ 0 ∧ f n ≠ 1 := by
+    by_contra! h
+    apply hf_nontriv
+    apply eq_on_Nat_iff_eq.1
+    intro n
+    by_cases hn0 : n=0
+    · simp only [hn0, CharP.cast_eq_zero, map_zero]
+    · push_neg at hn0
+      simp only [MulRingNorm.apply_one, Nat.cast_eq_zero, hn0, ↓reduceIte, h n hn0]
+  obtain ⟨n, hn1, hn2⟩ := hn
+  set P := {m : ℕ | 0 < f ↑m ∧ f ↑m < 1} -- p is going to be the minimum of this set
+  have hP : Set.Nonempty P := by
+    use n
+    exact ⟨map_pos_of_ne_zero f (Nat.cast_ne_zero.mpr hn1), lt_of_le_of_ne (bdd n) hn2⟩
+  use sInf P
+  exact ⟨Nat.sInf_mem hP, fun m ↦ fun hm ↦ Nat.sInf_le hm⟩
+
+-- ## Step 2: p is prime
+
+private lemma one_lt_of_ne_zero_one {a : ℕ} (ne_0 : a ≠ 0) (ne_1 : a ≠ 1) : 1 < a := by
+  rcases a with _ | a
+  · exact (ne_0 rfl).elim
+  · rw [Nat.succ_ne_succ, ← pos_iff_ne_zero] at ne_1
+    exact Nat.succ_lt_succ ne_1
+
+variable (p : ℕ) (hp0 : 0 < f p) (hp1 : f p < 1) (hmin : ∀ (m : ℕ), 0 < f m ∧ f m < 1 → p ≤ m)
+
+ /-- The minimal positive integer with absolute value smaller than 1 is a prime number-/
+lemma is_prime_of_smallest_Nat_zero_lt_mulRingNorm_lt_one : Prime p := by
+  rw [← Nat.irreducible_iff_prime]
+  constructor -- Two goals: p is not a unit and any product giving p must contain a unit
+  · rw [Nat.isUnit_iff]
+    intro p1
+    simp only [p1, Nat.cast_one, map_one, lt_self_iff_false] at hp1
+  · intro a b hab
+    rw [Nat.isUnit_iff, Nat.isUnit_iff]
+    by_contra! con
+    obtain ⟨a_neq_1, b_neq_1⟩ := con
+    apply not_le_of_lt hp1
+    /- GOAL: If p = a * b and a and b are both different form one, f p will be at least 1 giving
+    a contradiction -/
+    rw [hab, Nat.cast_mul, map_mul]
+    have hneq_0 {a b : ℕ} (hab : p = a * b) : a ≠ 0 := by
+      intro ha0
+      simp only [hab, ha0, zero_mul, CharP.cast_eq_zero, map_zero, lt_self_iff_false] at hp0
+    have one_le_f (a b : ℕ) (hab : p = a * b) (one_lt_b : 1 < b) : 1 ≤ f a := by
+      by_contra! hfa
+      apply not_le_of_gt (a:=p) (b:=a)
+      · rw [hab, gt_iff_lt]
+        exact lt_mul_of_one_lt_right (pos_iff_ne_zero.2 (hneq_0 hab)) one_lt_b
+      · apply hmin
+        refine ⟨?_, hfa⟩
+        apply map_pos_of_ne_zero
+        exact_mod_cast hneq_0 hab
+    have hba : p = b * a := by
+      rw [mul_comm]
+      exact hab
+    apply one_le_mul_of_one_le_of_one_le
+    · exact one_le_f a b hab (one_lt_of_ne_zero_one (hneq_0 hba) b_neq_1)
+    · exact one_le_f b a hba (one_lt_of_ne_zero_one (hneq_0 hab) a_neq_1)
+
+-- ## Step 3: if p does not divide m, then |m|=1
+
+open Real
+
+/-- A natural number not divible by p has absolute value 1 -/
+lemma mulRingNorm_eq_one_of_not_dvd (m : ℕ) (hpm : ¬ p ∣ m) : f m = 1 := by
```

Comment:
```suggestion
lemma mulRingNorm_eq_one_of_not_dvd {m : ℕ} (hpm : ¬ p ∣ m) : f m = 1 := by
```
`m` can be implicit, as it can be deduced from `hpm`. (Then when the lemma is used, the corresponding argument needs to be left out.)